pkg/cluster/updater.go

```diff
@@ -0,0 +1,372 @@
+		// Wait for new DS Pod.
+		for {
+			glog.Info("Waiting for new pod to come up...")
+			upl, err := client.Core().Pods(api.NamespaceSystem).List(lo)
+			if err != nil {
+				// Do not return an error here, since we have just deleted a pod,
+				// it is quite possible we have deleted an API Server pod. We may fail
+				// to contact it for a bit until the checkpointer API Server takes over.
+				glog.Error(err)
+			} else {
+				if ds.Status.DesiredNumberScheduled == int32(len(upl.Items)) {
+					if allContainersRunning(upl) {
+						if name == "kube-apiserver" {
+							// Make sure the apiserver is back online before we continue.
+							re := regexp.MustCompile("v[0-9].[0-9].[0-9]")
+							for {
+								glog.Info("waiting until correct API Server is running")
+								ver, err := client.Discovery().ServerVersion()
+								if err == nil {
+									cv := re.FindString(ver.GitVersion)
+									ev := re.FindString(image.Tag)
+									if cv != "" && cv == ev {
+										break
+									}
+									glog.Infof("got version %s, wanted %s", cv, ev)
+								}
+								time.Sleep(time.Second)
+							}
+						}
+						glog.Info("Done waiting for new pod")
+						break
+					}
+				}
+			}
+			time.Sleep(time.Second)
+		}
+	}
+	return nil
+}
+
+// DeploymentRollingUpdate will perform a safe rolling update on the Deployment
+// specified by `name`. An error will be returned if the rolling update failed.
+func DeploymentRollingUpdate(client clientset.Interface, name string, image *ContainerImage) error {
+	// Get the current deployment.
+	dep, err := client.Extensions().Deployments(api.NamespaceSystem).Get(name)
+	if err != nil {
+		return err
+	}
+	// Update the image in the specific container we care about (should match name
+	// of the deployment itself, per convention).
+	dep.Labels["version"] = image.Tag
+	dep.Spec.Template.Labels["version"] = image.Tag
+	for i, c := range dep.Spec.Template.Spec.Containers {
+		if c.Name == name {
+			dep.Spec.Template.Spec.Containers[i].Image = image.String()
+			break
+		}
+	}
+	// Update the deployment, which will trigger an update.
+	_, err = client.Extensions().Deployments(api.NamespaceSystem).Update(dep)
+	if err != nil {
+		return err
+	}
+	// Wait for update to finish.
+	l, err := labels.Parse(fmt.Sprintf("version=%s,k8s-app=%s", image.Tag, name))
+	if err != nil {
+		return err
+	}
+	lo := api.ListOptions{
+		LabelSelector: l,
+	}
+	replicas := int32(1)
+	if dep.Spec.Replicas != nil {
+		replicas = *dep.Spec.Replicas
+	}
+	for {
+		glog.Info("checking that all new containers are up")
+		upl, err := client.Core().Pods(api.NamespaceSystem).List(lo)
+		if err != nil {
+			// TODO might be worth returning error here.
+			glog.Error(err)
+		} else {
+			if int32(len(upl.Items)) == replicas {
+				if allContainersRunning(upl) {
+					return nil
+				}
+			}
+			glog.Infof("not all new containers have started, expected %d got %d", dep.Status.Replicas, len(upl.Items))
+		}
+		time.Sleep(time.Second)
```

Comment:
We need some kind of release valve here - we don't want it to be possible just to sit in this state forever.